### PR TITLE
Include Reg into microk8s

### PR DESCRIPTION
**Pull request by etheleon**
As suggested in this [thread](https://github.com/ubuntu/microk8s/issues/382#issuecomment-488608144), [reg](https://github.com/genuinetools/reg) will be a good addition to enable users to easily interact with local registry

### CODE REVIEW

**Comment by etheleon** [comment]
@ktsakalozos , the tests are failing any idea why?

File: microk8s-resources/actions/enable.reg.sh [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env bash
+
+set -e
+
+source $SNAP/actions/common/utils.sh
+
+read -ra ARGUMENTS <<< "$1"
+argz=("${ARGUMENTS[@]/#/--}")
```
Comment (ktsakalozos) [inline microk8s-resources/actions/enable.reg.sh]: I do not think you are doing anything with `argz` so lets simplify the code by removing these two lines.

File: microk8s-resources/actions/disable.reg.sh [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
+
+set -e
+
+source $SNAP/actions/common/utils.sh
+
+echo "Disabling Reg"
```
Comment (ktsakalozos) [inline microk8s-resources/actions/disable.reg.sh]: I think a single message should be enough. :)

File: microk8s-resources/actions/enable.reg.sh [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+# check if reg cli is already in the system.  Download if it doesn't exist.
+if [ ! -f "${SNAP_DATA}/bin/reg" ]; then
+  REG_VERSION="${REG_VERSION:-v0.16.0}"
+  echo "Fetching reg version $REG_VERSION."
+  sudo mkdir -p "$SNAP_DATA/bin"
+  sudo "${SNAP}/usr/bin/curl" -L https://github.com/genuinetools/reg/releases/download/${REG_VERSION}/reg-linux-amd64 -o "$SNAP_DATA/bin/reg"
```
Comment (ktsakalozos) [inline microk8s-resources/actions/enable.reg.sh]: MicroK8s ships for amd64 adn arm64, so we should use the  $SNAP_ARCH to get the right binary from the releases. Something like this should work: `https://github.com/genuinetools/reg/releases/download/${REG_VERSION}/reg-linux-${SNAP_ARCH}`.

**Comment by ktsakalozos** [comment]
@etheleon thank you for this PR. I left a few comments. It would be great if you could somehow add a test for this addon. Here is how to do that: 
- In https://github.com/ubuntu/microk8s/blob/master/tests/test-addons.py#L65 and https://github.com/ubuntu/microk8s/blob/master/tests/test-upgrade.py#L92 enable `reg` along with the registry.
- In https://github.com/ubuntu/microk8s/blob/master/tests/validators.py#L163 use the `reg` binary to do any operation on the private registry. An ls should return the my-busybox image, right?

If you follow the instruction and build MicroK8s from source https://github.com/ubuntu/microk8s#building-from-source you should see that you do not have a `microk8s.reg` command in your path. The reason is you are missing an entry into `snapcraft.yaml` like this one: https://github.com/ubuntu/microk8s/blob/master/snapcraft.yaml#L66 and the corresponding wrapper script like this one: https://github.com/ubuntu/microk8s/blob/master/microk8s-resources/wrappers/microk8s-linkerd.wrapper

It would have been better if the fist time you try the `microk8s.req` command the wrapper would download the binary then. This way you would not need the `microk8s.enable reg` step. What do you think?

File: microk8s-resources/actions/enable.reg.sh (hunk 1 above)
Comment (etheleon) [reply microk8s-resources/actions/enable.reg.sh]: Alright!

File: microk8s-resources/actions/disable.reg.sh (hunk 2 above)
Comment (etheleon) [reply microk8s-resources/actions/disable.reg.sh]: Alright!

File: microk8s-resources/actions/enable.reg.sh (hunk 3 above)
Comment (etheleon) [reply microk8s-resources/actions/enable.reg.sh]: Alright!

**Comment by etheleon** [comment]
@ktsakalozos Thanks for the detailed write up on writing the test! Let me get started on it! :)

**Comment by etheleon** [comment]
- snapcraft: add reg as app
- validator: include validator for `reg`
  * tags busybox
  * push to local registry
  * assets presence of image
  * deletes tag
- microk8s.reg wrapper: installs reg if binary not found
- microk8s.enable reg: installs reg (might be duplicate)

**Comment by ktsakalozos** [comment]
@etheleon are you able to see this: https://travis-ci.org/ubuntu/microk8s/builds/534665856?utm_source=github_status&utm_medium=notification  ?

I believe you are missing an import. 
```
       validate_reg()
      NameError: global name 'validate_reg' is not defined
```

**Comment by ktsakalozos** [comment]
@etheleon now we have:
```
>       output = (subprocess
                  .run(['reg', 'ls', '-f', 'localhost:32000'], stdout=subprocess.PIPE)
                  .stdout
                  .decode('utf-8')
                  .split("\n"))
E       AttributeError: 'module' object has no attribute 'run'
```

I should document how to run the tests locally, (they are run with pytest).

File: tests/test-upgrade.py [hunk 4]
```diff
@@ -96,6 +97,15 @@ def test_upgrade(self):
         except:
             print('Will not test registry')
 
+
+        try:
+            enable = microk8s_enable("reg")
```
Comment (ktsakalozos) [inline tests/test-upgrade.py]: Can you please add a `timeout_insec=30` when enabling the add-on. If we do not set a timeout enabling tries for too long and tests on travis timeout.

**Comment by ktsakalozos** [comment]
Closing this PR due to inactivity.

**Pull request closed without merging**